File: apps/mobile/lib/services/home_widget_sync.dart
import 'package:home_widget/home_widget.dart';

/// Pushes data to the native home-screen widgets (iOS WidgetKit today; the
/// same data layer works for an Android widget later).
///
/// The App Group id, the data key, and the iOS widget "kind" must all stay in
/// sync with `ios/StreakWidget/StreakWidget.swift` and the App Group capability
/// on the Runner + extension targets.
class HomeWidgetSync {
  static const _appGroupId = 'group.com.heftor.heftor';
  static const _streakKey = 'streak';

  // Matches the WidgetKit `kind` string in StreakWidget.swift.
  static const _iosWidgetKind = 'StreakWidget';

  /// Call once at startup so every later write lands in the shared container.
  static Future<void> init() async {
    try {
      await HomeWidget.setAppGroupId(_appGroupId);
    } catch (_) {
      // Home widgets are a nice-to-have — never block app startup.
    }
  }

  /// Persist the current weekly streak and ask the OS to redraw the widget.
  /// Best-effort: failures (e.g. no widget added yet) are swallowed.
  static Future<void> updateStreak(int streak) async {
    try {
      await HomeWidget.setAppGroupId(_appGroupId);
      await HomeWidget.saveWidgetData<int>(_streakKey, streak);
      await HomeWidget.updateWidget(iOSName: _iosWidgetKind);
    } catch (_) {
      // Ignore — the widget simply keeps its last value.
    }
  }
}

File: apps/native-mobile/src/lib/homeWidgetSync.ts
// Ported 1:1 from apps/mobile/lib/services/home_widget_sync.dart.
//
// NO-OP STUB. The Flutter original pushed streak / current-month / this-week
// data into a shared App Group container that a native iOS WidgetKit widget
// (and a future Android widget) read off the home screen.
//
// Expo's managed workflow has no first-class home-screen widget: shipping one
// needs a native config plugin + a custom dev build (e.g. @bacons/apple-targets,
// or a bare workflow with a WidgetKit extension + App Group entitlement). That
// work is deferred, so these methods keep the same names/shape as the Dart
// original but are safe no-ops. Callers can wire them up now; the data will
// start flowing once the native widget lands.
export class HomeWidgetSync {
  /// Call once at startup. No-op until a native home widget exists.
  static async init(): Promise<void> {
    // no-op — see file header.
  }

  /// Persist everything the home-screen widgets need and ask the OS to redraw.
  /// Mirrors the Dart `sync({streak, workoutDates, weekGoal})` surface.
  ///
  /// [workoutDates] is the full list of dates with a completed workout;
  /// [weekGoal] is the user's target training days per week.
  static async sync(_opts: {
    streak: number;
    workoutDates: Date[];
    weekGoal: number;
  }): Promise<void> {
    // no-op — see file header.
  }
}
